src/Juvix/Compiler/Core/Evaluator.hs

```diff
@@ -55,7 +55,9 @@ eval !ctx !env0 = convertRuntimeNodes . eval' env0
 
     eval' :: Env -> Node -> Node
     eval' !env !n = case n of
-      NVar (Var _ idx) -> env !! idx
+      NVar (Var _ idx)
+        | idx < length env -> env !! idx
+        | otherwise -> error ("invalid index: " <> show idx <> " (length: " <> show (length env) <> ")")
```

Comment:
No! We don't want to compute `length env` (in O(number of binders)) on every variable access in the evaluator! If you really want to have bounds checking then reimplement `!!`.

Reply:
I have updated it to use `drop`. Is it fine now?

Reply:
Actually, I think it will be better to use https://hackage.haskell.org/package/strict-0.4.0.1/docs/Data-Strict-Maybe.html, so the `listToMaybe` does not introduce an unnecessary thunk

Reply:
let's keep it as it was before and use the Prelude's `!!`. Now that we have the scope checker, this error message is not as important. And we can always add it back temporarily when debugging.